docs/metrics.md

```diff
@@ -0,0 +1,561 @@
+```
+
+This approach minimizes memory allocations on metrics hot paths, improving
+overall performance when reporting metrics frequently.
+
+### Reporting measurements via asynchronous instruments
+
+Asynchronous instruments like `ObservableCounter` are ideal for reporting
+metrics that are already being tracked or stored elsewhere in your application.
+These instruments allow you to observe and report the current state of existing
+values without manually managing the aggregation yourself.
+
+:heavy_check_mark: Use `ObservableCounter` when you already have a variable
+tracking a count:
+
+```rust
+// Example: Using ObservableCounter when you already have a variable tracking counts
+use opentelemetry::KeyValue;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+// An existing variable in your application
+static REQUEST_COUNT: AtomicU64 = AtomicU64::new(0);
+
+// In your application code, you update this directly
+fn handle_request() {
+    // Process request...
+    REQUEST_COUNT.fetch_add(1, Ordering::SeqCst);
+}
+
+// When setting up metrics, register an observable counter that reads from your variable
+fn setup_metrics(meter: &opentelemetry::metrics::Meter) {
+    let _observable_counter = meter
+        .u64_observable_counter("request_count")
+        .with_description("Number of requests processed")
+        .with_unit("requests")
+        .with_callback(|observer| {
+            // Read the current value from your existing counter
+            observer.observe(
+                REQUEST_COUNT.load(Ordering::SeqCst), 
+                &[KeyValue::new("endpoint", "/api")]
+            )
+        })
+        .build();
+}
+```
+
+:heavy_check_mark: Use regular Counter when you need to increment counts at
+specific points in your code:
+
+```rust
+// Example: Using Counter when incrementing at specific code points
+use opentelemetry::KeyValue;
+
+fn setup_metrics(meter: &opentelemetry::metrics::Meter) -> opentelemetry::metrics::Counter<u64> {
+    meter
+        .u64_counter("processed_items")
+        .with_description("Number of items processed")
+        .build()
+}
+
+// Then in your processing code:
+fn process_item(counter: &opentelemetry::metrics::Counter<u64>, item_type: &str) {
+    // Process item...
+    
+    // Increment the counter with the item type as an attribute
+    counter.add(1, &[KeyValue::new("type", item_type)]);
+}
+```
+
+## MeterProvider Management
+
+:stop_sign: Avoid creating `MeterProvider` instances too frequently. A
+`MeterProvider` is resource-intensive and designed to be reused throughout the
+application. Typically, one `MeterProvider` instance per process is sufficient.
+
+:heavy_check_mark: Properly manage the lifecycle of `MeterProvider` instances if
+you create them. Follow these guidelines:
+
+* **Create Once, Reuse Always**: Create a single `MeterProvider` instance and
+  reuse it throughout the application. Avoid creating multiple instances to
+  conserve resources.
```

Comment:
This part seems redundant since the same advice is already given under the stop sign warning.